app/lib/shared/configuration.dart

```diff
@@ -46,63 +57,78 @@ class Configuration {
   /// the Datastore.
-  auth.ServiceAccountCredentials get credentials {
-    if (_credentials == null) {
-      _credentials = new auth.ServiceAccountCredentials.fromJson(
-          new File(envConfig.gcloudKey).readAsStringSync());
-    }
-    return _credentials;
-  }
-
-  bool get hasCredentials => envConfig.hasCredentials;
+  final auth.ServiceAccountCredentials credentials;
 
   /// Create a configuration for production deployment.
   ///
   /// This will use the Datastore from the cloud project and the Cloud Storage
   /// bucket 'pub-packages'. The credentials for accessing the Cloud
   /// Storage is retrieved from the Datastore.
-  Configuration._prod() : this._local('dartlang-pub');
+  factory Configuration.prod() {
+    final projectId = 'dartlang-pub';
+    return Configuration(
+        projectId: projectId,
+        packageBucketName: 'pub-packages',
+        dartdocStorageBucketName: '$projectId--dartdoc-storage',
+        popularityDumpBucketName: '$projectId--popularity',
+        searchSnapshotBucketName: '$projectId--search-snapshot',
+        backupSnapshotBucketName: '$projectId--backup-snapshots',
+        searchServicePrefix: 'https://search-dot-$projectId.appspot.com',
+        oauthRedirectUrl: 'https://pub.dartlang.org/oauth/callback',
+        pubClientAudience: _pubClientAudience,
+        pubSiteAudience:
+            '818368855108-e8skaopm5ih5nbb82vhh66k7ft5o7dn3.apps.googleusercontent.com',
+        credentials: _loadCredentials());
+  }
 
   /// Create a configuration for development/staging deployment.
-  Configuration._dev() : this._local('dartlang-pub-dev');
-
-  /// Base configuration for local development
-  Configuration._local(String projectId)
-      : projectId = projectId,
-        packageBucketName = projectId == 'dartlang-pub'
-            ? 'pub-packages'
-            : '$projectId--pub-packages',
-        analyzerServicePrefix = 'https://analyzer-dot-$projectId.appspot.com',
-        dartdocServicePrefix = 'https://dartdoc-dot-$projectId.appspot.com',
-        searchServicePrefix = 'https://search-dot-$projectId.appspot.com',
-        dartdocStorageBucketName = '$projectId--dartdoc-storage',
-        popularityDumpBucketName = '$projectId--popularity',
-        searchSnapshotBucketName = '$projectId--search-snapshot',
-        backupSnapshotBucketName = '$projectId--backup-snapshots',
-        oauthRedirectUrl = projectId == 'dartlang-pub'
-            ? 'https://pub.dartlang.org/oauth/callback'
-            : 'https://dartlang-pub-dev.appspot.com/oauth/callback',
-        pubSiteAudience = projectId == 'dartlang-pub'
-            ? '818368855108-e8skaopm5ih5nbb82vhh66k7ft5o7dn3.apps.googleusercontent.com'
-            : '621485135717-idb8t8nnguphtu2drfn2u4ig7r56rm6n.apps.googleusercontent.com';
+  factory Configuration.dev() {
+    final projectId = 'dartlang-pub-dev';
+    return Configuration(
+      projectId: projectId,
+      packageBucketName: '$projectId--pub-packages',
```

Comment:
**Unrelated:** but analysis and dartdoc still use prod bucket for packages on staging, right?
(we probably should change that, such that analysis and dartdoc uses the staging server by specifying `PUB_HOSTED_URL`, right?)

This just got me thinking that we would be able to test a lot more if we did that :)